logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -294,12 +310,58 @@ private boolean isOutput(final Vertex vertex) {
         return outputs.containsKey(vertex.getId());
     }
 
+    public final class CompiledOrderedExecution extends CompiledExecution {
+
+        @Override
+        public void compute(final QueueBatch batch, final boolean flush, final boolean shutdown) {
+           compute(batch.collection(), flush, shutdown);
+        }
+
+        @SuppressWarnings({"rawtypes"})
+        @Override
+        public void compute(final RubyArray batch, final boolean flush, final boolean shutdown) {
+            compute(batch, flush, shutdown);
+        }
+
+        @SuppressWarnings({"rawtypes", "unchecked"})
+        private void compute(final Collection batch, final boolean flush, final boolean shutdown) {
+            final RubyArray<RubyEvent> outputBatch = RubyUtil.RUBY.newArray();
+            // send batch one-by-one as single-element batches down the filters
+            final RubyArray<RubyEvent> filterBatch = RubyUtil.RUBY.newArray(1);
+            batch.forEach((e) -> {
+                filterBatch.set(0, e);
+                final Collection<RubyEvent> result = compiledFilters.compute(filterBatch, flush, shutdown);
+                copyNonCancelledEvents(result, outputBatch);
+                compiledFilters.clear();
+            });
+            compiledOutputs.compute(outputBatch, flush, shutdown);
+        }
+    }
+
+    public final class CompiledUnorderedExecution extends CompiledExecution {
+
+        @Override
+        public void compute(final QueueBatch batch, final boolean flush, final boolean shutdown) {
+            compute(batch.to_a(), flush, shutdown);
```

Comment:
🤔 so we end up needing this arrayification anyway, which appears to be a bit of a bummer.

Reply:
yeah, that't because the `compute` interface https://github.com/elastic/logstash/blob/aa931661c6773052e899b7c9fb7ee2d76ececcf2/logstash-core/src/main/java/org/logstash/config/ir/compiler/Dataset.java#L67 
takes a `RubyArray` as param. 
I will look into what could be done with that, these array conversions are very costly, I suggest we defer that into another issue though.

Reply:
+1 to deferring